### PR TITLE
Feature project report detail

### PR DESCRIPTION
**Pull request by maciejSamerdak**
Introduces report editing view linked to **project-report-list**.

Related #133

### CODE REVIEW

**Comment by maciejSamerdak** [comment]
Due to lack of admin's serializer, I changed this view to a Django-based generic view, in resemblance to AdminReportView.

File: employees/tests/test_unit_report_viewset.py [hunk 1]
```diff
@@ -642,3 +643,100 @@ def test_project_report_list_get_queryset_method_should_return_queryset_containi
...
+
+    def test_project_report_detail_view_should_display_report_details(self):
+        request = APIRequestFactory().get(path=reverse("project-report-detail", args=(self.report.pk,)))
+        request.user = self.user
+        response = ProjectReportDetail.as_view()(request, pk=self.report.pk)
+        self.assertEqual(response.status_code, 200)
+        self.assertContains(response, self.report.description)
+        self.assertEqual(response.data["serializer"].instance, self.report)
+
+    def test_project_report_detail_view_should_not_be_accessible_for_unauthenticated_users(self):
```
Comment (rwrzesien) [inline employees/tests/test_unit_report_viewset.py]: I think this test is not needed, `login_decorator` is working well :) Lets just test our custom logic.

File: employees/tests/test_unit_report_viewset.py [hunk 2]
```diff
@@ -642,3 +643,100 @@ def test_project_report_list_get_queryset_method_should_return_queryset_containi
...
+        request = APIRequestFactory().get(path=reverse("project-report-detail", args=(self.report.pk,)))
+        request.user = AnonymousUser()
+        response = ProjectReportDetail.as_view()(request, pk=self.report.pk)
+        self.assertEqual(response.status_code, 403)
+
+    def test_project_report_detail_view_should_not_render_non_existing_report(self):
```
Comment (rwrzesien) [inline employees/tests/test_unit_report_viewset.py]: Same here, routing should work OK by default too, I think this test is not necessary.

File: employees/tests/test_unit_report_viewset.py [hunk 3]
```diff
@@ -642,3 +643,100 @@ def test_project_report_list_get_queryset_method_should_return_queryset_containi
         self.assertFalse(other_project_report in queryset)
         self.assertEqual(queryset[0], other_report_1)
         self.assertEqual(queryset[1], self.report)
+
+
+class ProjectReportDetailTests(TestCase):
+    def setUp(self):
+        self.user = CustomUser(
```
Comment (rwrzesien) [inline employees/tests/test_unit_report_viewset.py]: Please use `UserFactory`.

File: employees/tests/test_unit_report_viewset.py [hunk 4]
```diff
@@ -642,3 +643,100 @@ def test_project_report_list_get_queryset_method_should_return_queryset_containi
...
+            email="testuser@codepoets.it", password="newuserpasswd", first_name="John", last_name="Doe", country="PL"
+        )
+        self.user.full_clean()
+        self.user.save()
+
+        self.project = Project(name="Test Project", start_date=datetime.datetime.now())
+        self.project.full_clean()
+        self.project.save()
+
+        self.report = Report(
```
Comment (rwrzesien) [inline employees/tests/test_unit_report_viewset.py]: Please user `ReportFactory`. Then you don't have to create `self.project` too.

File: employees/templates/employees/project_report_detail.html [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+{% block content %}
+<h1>{{ UI_text.PAGE_TITLE.value }}{{ report.project }} ({{ report.date }})</h1>
+</br>
+<div class="modal-dialog">
+    <form action="{% url 'project-report-detail' pk=report.pk %}" method="POST">
+        {% csrf_token %}
+        {{ form|crispy }}
+        <input type="submit" value="{{ UI_text.UPDATE_REPORT_BUTTON.value }}">
+        <input type="submit" name="discard" value="{{ UI_text.DISCARD_CHANGES_BUTTON.value }}">
```
Comment (rwrzesien) [inline employees/templates/employees/project_report_detail.html]: Is this being handled by Django by default? 

Generally I don't like making a POST request just to **not** make any changes. I think from UX perspective it is simpler to "go back" or "cancel" or something like that. I would suggest replacing it with a link back.

File: employees/tests/test_unit_report_viewset.py [hunk 6]
```diff
@@ -642,3 +643,100 @@ def test_project_report_list_get_queryset_method_should_return_queryset_containi
...
+            description="Some description",
+            author=self.user,
+            project=self.project,
+            work_hours=Decimal("8.00"),
+        )
+        self.report.full_clean()
+        self.report.save()
+
+    def test_project_report_detail_view_should_display_report_details(self):
+        request = APIRequestFactory().get(path=reverse("project-report-detail", args=(self.report.pk,)))
```
Comment (rwrzesien) [inline employees/tests/test_unit_report_viewset.py]: Please use `self.client` that is delivered with Django `TestCase`. You can simply make `self.client.force_login(self.user)` to make it authenticated (ie. in `setUp()`). You can take a look how it is used in other refactored tests for Django views.

File: employees/tests/test_unit_report_viewset.py [hunk 7]
```diff
@@ -642,3 +643,100 @@ def test_project_report_list_get_queryset_method_should_return_queryset_containi
...
+        self.assertEqual(response.status_code, 302)
+        self.assertEqual(current_description, new_description)
+        self.assertTrue(self.report.editable)
+
+    def test_project_report_detail_view_should_not_update_report_on_discard(self):
+        new_description = "Some other description"
+        request = APIRequestFactory().post(
+            path=reverse("project-report-detail", args=(self.report.pk,)),
+            data={
+                "date": datetime.datetime.now().date(),
```
Comment (rwrzesien) [inline employees/tests/test_unit_report_viewset.py]: If its reused (or part of it is), add it to `setUp()` like `self.data`.

**Comment by rwrzesien** [comment]
@maciejSamerdak Looks nice, please just adapt the test and please explain me how the "discard" works :)

File: employees/tests/test_unit_report_viewset.py (hunk 1 above)
Comment (maciejSamerdak) [reply employees/tests/test_unit_report_viewset.py]: You think so? This way at least we know that the view isn't protected anymore if anything goes wrong. Or are there other, better ways to test that?

File: employees/templates/employees/project_report_detail.html (hunk 5 above)
Comment (maciejSamerdak) [reply employees/templates/employees/project_report_detail.html]: I'm happy to inform you that everything is in absolute order and this solution in fact does not work at all. 😄 

Not only does the test fail, I also cannot see any other button except for **Update** in rendered view. I guess I must have skipped the unit tests on this branch, during the hassle of multiple rebases.

Comment (maciejSamerdak) [reply employees/templates/employees/project_report_detail.html]: Well, I'm happy to inform you that it does not work at all. 😁 

The _discard_ button just works as another submit, changes apply. I'll fix it up in a jiff.

**Comment by maciejSamerdak** [comment]
@rwrzesien I added requested changes. I also refactored unit tests a little and moved them to `test_views`. However, I'd like to highlight one issue: after refactoring the tests, they performance dropped significantly. It's like a double-action revolver vs a Tommy gun.

File: employees/tests/test_unit_report_viewset.py (hunk 2 above)
Comment (maciejSamerdak) [reply employees/tests/test_unit_report_viewset.py]: Done

File: employees/tests/test_unit_report_viewset.py (hunk 3 above)
Comment (maciejSamerdak) [reply employees/tests/test_unit_report_viewset.py]: Done

File: employees/tests/test_unit_report_viewset.py (hunk 4 above)
Comment (maciejSamerdak) [reply employees/tests/test_unit_report_viewset.py]: Done

File: employees/tests/test_unit_report_viewset.py (hunk 6 above)
Comment (maciejSamerdak) [reply employees/tests/test_unit_report_viewset.py]: Done

File: employees/tests/test_unit_report_viewset.py (hunk 7 above)
Comment (maciejSamerdak) [reply employees/tests/test_unit_report_viewset.py]: Done

File: employees/templates/employees/project_report_detail.html (hunk 5 above)
Comment (maciejSamerdak) [reply employees/templates/employees/project_report_detail.html]: Done

File: employees/tests/test_unit_report_viewset.py (hunk 1 above)
Comment (rwrzesien) [reply employees/tests/test_unit_report_viewset.py]: Right, lets keep it then.

**Comment by rwrzesien** [comment]
@maciejSamerdak 
> It's like a double-action revolver vs a Tommy gun.

Nice comparison :) Well that is because previously we was running view directly as a function. Now it is sending request through all other stuff like middleware, routing etc. - basically in the same way as you would send a request to Django server running. It takes slower **but** also behaves more like a regular HTTP call, allowing us to test/discover more stuff (for example if any other layer has impact on the view).

**Comment by rwrzesien** [comment]
@maciejSamerdak Looks good, tests are much more compact now, this is exactly what I have meant. If you want add back the test `test_project_report_detail_view_should_not_be_accessible_for_unauthenticated_users` but if you don't want its fine for me too.

**Comment by maciejSamerdak** [comment]
I recovered the login-required test, just in case. When integration tests are introduced, covering this mechanism in greater detail, it could be removed.

**Comment by kbeker** [comment]
Accidentally closed. Merged

**Pull request closed without merging**